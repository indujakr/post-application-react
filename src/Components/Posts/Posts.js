import React, { useContext, useState, useRef } from "react";
import { makeStyles } from "@material-ui/core/styles";
import { red } from "@material-ui/core/colors";
import FavoriteIcon from "@material-ui/icons/Favorite";
import FavoriteBorderOutlinedIcon from "@material-ui/icons/FavoriteBorderOutlined";
import EditIcon from "@material-ui/icons/Edit";
import DeleteIcon from "@material-ui/icons/Delete";
import CheckIcon from "@material-ui/icons/Check";
import {
  Avatar,
  Card,
  CardHeader,
  CardContent,
  CardActions,
  IconButton,
  Typography,
  Input,
} from "@material-ui/core";
import { StoreDataProvider } from "../../Store/Store";
import axios from "axios";
import TextareaAutosize from "@mui/base/TextareaAutosize";
import SnackBarAlert from "../../Common/SnackBarAlert";

const useStyles = makeStyles(() => ({
  root: {
    marginTop: 16,
  },
  alignText: {
    textAlign: "left",
  },
  avatar: {
    backgroundColor: red[500],
  },
}));

const Posts = ({ postDetails, edit, setEdit }) => {
  const classes = useStyles();
  const [success, setSuccess] = useState({});
  const title = useRef(postDetails.title);
  const body = useRef(postDetails.body);
  const store = useContext(StoreDataProvider);

  /**
   * deleteHandler deletes the selected post
   * @param id
   */
  const deleteHandler = (id) => {
    axios
      .delete(`https://jsonplaceholder.typicode.com/posts/${id}`)
      .then((response) => {
        if (Object.keys(response.data).length === 0) {
          setSuccess([true, "Post deleted successfully!", "success"]);
        }
      })
      .catch((err) => {
        setSuccess([true, err, "error"]);
      });
  };

  /**
   * editHandler updated the selected post
   * @param id
   */
  const editHandler = (id) => {
    if (edit === 0) {
      setEdit(postDetails.id);
    } else if (edit === id) {
      if (title.current?.length > 0 && body.current?.length > 0) {
        if (
          title.current === postDetails.title &&
          body.current === postDetails.body
        ) {
          setEdit(0);
          setSuccess([true, "No changes to save!", "warning"]);
        } else {
          axios
            .patch(`https://jsonplaceholder.typicode.com/posts/${id}`, {
              title: title.current,
              body: body.current,
            })
            .then((response) => {
              setEdit(0);
              setSuccess([true, "Post updated successfully!", "success"]);
            })
            .catch((err) => {
              setSuccess([true, err, "error"]);
            });
        }
      } else {
        if (title.current?.length === 0 && body.current?.length === 0)
          setSuccess([true, "Please Input Post Title & Content!", "error"]);
        else if (title.current?.length === 0)
          setSuccess([true, "Please Input Title Content!", "error"]);
        else if (body.current?.length === 0)
          setSuccess([true, "Please Input Post Content!", "error"]);
      }
    } else {
      setSuccess([true, "Please update one at a time", "error"]);
    }
  };

  /**
   * loadTitleHandler loads the title based on read / edit mode 
   */
  const loadTitleHandler = () => {
    return edit === postDetails.id ? (
      <Input
        fullWidth
        className={classes.alignText}
        defaultValue={postDetails.title}
        onBlur={(event) => {
          title.current = event.target.value;
        }}
      />
    ) : (
      postDetails.title
    );
  };

  return (
    <>
      <Card className={classes.root} key={postDetails.id} data-testid="post-card">
        <CardHeader
          className={classes.alignText}
          title={loadTitleHandler()}
          subheader={
            postDetails.userId === 1
              ? `${postDetails.user} (me)`
              : postDetails.user
          }
          avatar={
            <Avatar style={{ background: `${postDetails.color}` }}>
              {postDetails.user.split(" ")[0][0]}
              {postDetails.user.split(" ")[1][0]}
            </Avatar>
          }
        />
        <CardContent>
          {edit === postDetails.id ? (
            <TextareaAutosize
              aria-label="minimum height"
              minRows={2}
              style={{ width: "100%" }}
              className={classes.alignText}
              defaultValue={postDetails.body}
              onBlur={(event) => {
                body.current = event.target.value;
              }}
            />
          ) : (
            <Typography
              variant="body2"
              color="textSecondary"
              className={classes.alignText}
            >
              {postDetails.body}
            </Typography>
          )}
        </CardContent>
        <CardActions disableSpacing>
          <IconButton
            onClick={() =>
              store.setPostContent({
                type: "UPDATE_LIKE",
                value: [postDetails.id, !postDetails.isLiked],
              })
            }
          >
            {postDetails.isLiked ? (
              <FavoriteIcon style={{ color: "#f50057" }} />
            ) : (
              <FavoriteBorderOutlinedIcon />
            )}
          </IconButton>
          {postDetails.userId === 1 && (
            <IconButton onClick={() => editHandler(postDetails.id)}>
              {edit === postDetails.id ? <CheckIcon /> : <EditIcon />}
            </IconButton>
          )}
          <IconButton onClick={() => deleteHandler(postDetails.id)}>
            <DeleteIcon />
          </IconButton>
        </CardActions>
      </Card>
      <SnackBarAlert success={success} setSuccess={setSuccess} />
    </>
  );
};

export default Posts;