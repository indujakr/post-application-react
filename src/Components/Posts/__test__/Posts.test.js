import React from "react";
import { render, screen } from '@testing-library/react';
import Posts from "../Posts";
import renderer from "react-test-renderer";

let postDetails ={
    userId: 1,
    id: 2,
    user:"Leanne Graham",
    color:"rgb(225, 174, 177)",
    title: "Title Content",
    body: "Body Content"
  }

test("render post card without crashing",()=>{
    render(<Posts postDetails={postDetails} />);
    const cardElement = screen.getByTestId('post-card');
    expect(cardElement).toBeInTheDocument();
})

test("render post card data with crashing",()=>{
    render(<Posts postDetails={postDetails} />);
    const titleElement = screen.getByText('Title Content');
    const bodyElement = screen.getByText('Body Content');
    expect(titleElement).toBeInTheDocument();
    expect(bodyElement).toBeInTheDocument();
})

test("match snapshot",()=>{
    const element = renderer.create(<Posts postDetails={postDetails} />).toJSON();
    expect(element).toMatchSnapshot();
})

