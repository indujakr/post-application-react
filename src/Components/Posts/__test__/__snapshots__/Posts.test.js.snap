// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match snapshot 1`] = `
<div
  className="MuiPaper-root MuiCard-root makeStyles-root-7 MuiPaper-elevation1 MuiPaper-rounded"
  data-testid="post-card"
>
  <div
    className="MuiCardHeader-root makeStyles-alignText-8"
  >
    <div
      className="MuiCardHeader-avatar"
    >
      <div
        className="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault"
        style={
          Object {
            "background": "rgb(225, 174, 177)",
          }
        }
      >
        L
        G
      </div>
    </div>
    <div
      className="MuiCardHeader-content"
    >
      <span
        className="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
      >
        Title Content
      </span>
      <span
        className="MuiTypography-root MuiCardHeader-subheader MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
      >
        Leanne Graham (me)
      </span>
    </div>
  </div>
  <div
    className="MuiCardContent-root"
  >
    <p
      className="MuiTypography-root makeStyles-alignText-8 MuiTypography-body2 MuiTypography-colorTextSecondary"
    >
      Body Content
    </p>
  </div>
  <div
    className="MuiCardActions-root"
  >
    <button
      className="MuiButtonBase-root MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55l-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
          />
        </svg>
      </span>
    </button>
    <button
      className="MuiButtonBase-root MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
          />
        </svg>
      </span>
    </button>
    <button
      className="MuiButtonBase-root MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;
